letsencrypt/models/letsencrypt.py

```diff
@@ -4,165 +4,135 @@
-            cfg.file.flush()
-            cmdline = [
-                'openssl', 'req', '-new',
-                parameter_model.get_param(
-                    'letsencrypt.openssl.digest', '-sha256'),
-                '-key', self.generate_domain_key(domain),
-                '-subj', '/CN=%s' % domain, '-config', cfg.name,
-                '-out', csr,
-            ]
-            if len(domains) > 1:
-                cmdline.extend([
-                    '-reqexts', 'SAN',
-                ])
-            self.call_cmdline(cmdline)
-        return csr
+                    _("Let's encrypt doesn't work with private addresses "
+                        "or local domains!"))
 
     @api.model
     def cron(self):
         domain = urlparse.urlparse(
-            self.env['ir.config_parameter'].get_param(
-                'web.base.url', 'localhost')).netloc
-        self.validate_domain(domain)
-        account_key = self.generate_account_key()
-        csr = self.generate_csr(domain)
-        acme_challenge = get_challenge_dir()
-        if not os.path.isdir(acme_challenge):
-            os.makedirs(acme_challenge)
-        if self.env.context.get('letsencrypt_dry_run'):
-            crt_text = 'I\'m a test text'
-        else:  # pragma: no cover
-            from acme_tiny import get_crt, DEFAULT_CA
-            crt_text = get_crt(
-                account_key, csr, acme_challenge, log=_logger, CA=DEFAULT_CA)
-        with open(os.path.join(get_data_dir(), '%s.crt' % domain), 'w')\
-                as crt:
-            crt.write(crt_text)
-            chain_cert = urllib2.urlopen(
                 self.env['ir.config_parameter'].get_param(
-                    'letsencrypt.chain_certificate_address',
-                    'https://letsencrypt.org/certs/'
-                    'lets-encrypt-x3-cross-signed.pem')
-            )
-            crt.write(chain_cert.read())
-            chain_cert.close()
-            _logger.info('wrote %s', crt.name)
-        reload_cmd = self.env['ir.config_parameter'].get_param(
-            'letsencrypt.reload_command', False)
-        if reload_cmd:
-            _logger.info('reloading webserver...')
-            self.call_cmdline(['sh', '-c', reload_cmd])
+                    'web.base.url', 'localhost')).netloc
+        self.validate_domain(domain)
+        # Generate account key
+        account_key_file = self._generate_key('account_key')
+        account_key = jose.JWKRSA.load(open(account_key_file).read())
+        # Generate domain key
+        domain_key_file = self._generate_key(domain)
+        net = client.ClientNetwork(account_key)
+        acme = client.ClientV2(
+            V2_STAGING_DIRECTORY_URL,
+            net)
+        _logger.info('Auto-accepting TOS')
+        # TODO if you try to register when you already have, this will crash.
+        acme.new_account(NewRegistration())
+        from acme import crypto_util
+        csr = crypto_util.make_csr(open(domain_key_file).read(), [domain])
```

Comment:
you dropped support for multiple alternative names, check the original csr creating code. I think you should slightly modify this, still support the naming scheme with numbered parameters, but do a `str.split()` on all of them - this way, you can just fill in a space separated list of domain names, much simpler to handle. i'm ashamed i didn't implement it like this in the first place. this then you can also expose in the config page